File: sorting/merge-sort.js
"use strict";
const utils = require("../utils");

var testSizes = [100, 1000, 10000, 100000, 1000000];
var testTimeResults = [];

testSizes.forEach(n => {
  var toBeSorted, sorted, startTime, endTime;
  toBeSorted = utils.createRandomArray(n);
  startTime = +new Date();

  sorted = mergeSort(toBeSorted);

  endTime = +new Date();
  console.log(`n=${n} took ${endTime - startTime}ms`);
  testTimeResults.push(endTime - startTime);

  // Check if the array is sorted
  utils.isSorted(sorted);
});

function merge(arr1, arr2) {
  var p1, p2, sortedArray;
  p1 = 0;
  p2 = 0;
  sortedArray = [];

  // Copy elements to sortedArray until one of the arrays is completed
  while (p1 < arr1.length && p2 < arr2.length) {
    if (arr1[p1] <= arr2[p2]) {
      sortedArray.push(arr1[p1]);
      p1++;
    } else {
      sortedArray.push(arr2[p2]);
      p2++;
    }
  }

  // Check if one of arrays was not completely copied to sortedArray
  while (p1 < arr1.length) {
    sortedArray.push(arr1[p1]);
    p1++;
  }
  while (p2 < arr2.length) {
    sortedArray.push(arr2[p2]);
    p2++;
  }
  return sortedArray;
}

function mergeSort(srcArray) {
  if (srcArray.length === 1 || srcArray.length === 0) return srcArray;
  var half, leftArray, rightArray;
  half = Math.floor(srcArray.length / 2);
  // split array
  leftArray = mergeSort(srcArray.slice(0, half));
  rightArray = mergeSort(srcArray.slice(half, srcArray.length));
  // merge sorted arrays
  return merge(leftArray, rightArray);
}

File: sorting/merge_sort.py
import random
import time

from utils.arrays import create_random_array, is_sorted


def merge(leftArray, rightArray):
    sorted_array = []
    lp, rp = 0, 0
    while (lp < len(leftArray) and rp < len(rightArray)):
        if (leftArray[lp] <= rightArray[rp]):
            sorted_array.append(leftArray[lp])
            lp += 1
        else:
            sorted_array.append(rightArray[rp])
            rp += 1

    while (lp < len(leftArray)):
        sorted_array.append(leftArray[lp])
        lp += 1

    while (rp < len(rightArray)):
        sorted_array.append(rightArray[rp])
        rp += 1

    return sorted_array


def mergesort(array):
    array_length = len(array)
    if (array_length == 0 or array_length == 1):
        return array

    half = array_length // 2
    left = mergesort(array[0:half])
    right = mergesort(array[half:array_length])

    return merge(left, right)


if __name__ == '__main__':
    size = 10000
    random_list = create_random_array(size)
    t0 = time.time()
    sorted_list = mergesort(random_list)
    t1 = time.time()
    if (not is_sorted(sorted_list)):
        raise Exception('The array was not correctly sorted.')
    if (size < 20):
        print('Unsorted array   =>    {}'.format(random_list))
        print('Sorted array     =>    {}'.format(sorted_list))
    print('Took {:.3f}ms to sort {} elements.'.format((t1 - t0) * 1000, size))
